---
title: "R Notebook"
output: html_notebook
---

```{r}
library(dada2)
library(tidyverse)
library(DECIPHER)

```


```{r}
fnFs <- sort(list.files("lane1/", pattern="_R1_001.fastq", full.names = TRUE))
fnRs <- sort(list.files("lane1", pattern="_R2_001.fastq", full.names = TRUE))

sample.names <- sapply(strsplit(basename(fnFs), "_S"), `[`, 1)
sample.names
```

```{r}
plotQualityProfile(fnFs[1:2])

```


```{r}
plotQualityProfile(fnRs[1:2])

```


```{r}
filtFs <- file.path("filtered_fastq", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path("filtered_fastq", paste0(sample.names, "_R_filt.fastq.gz"))
names(filtFs) <- sample.names
names(filtRs) <- sample.names

```

```{r}
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(240,160), maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE, compress=TRUE, multithread=TRUE)
head(out)

```

```{r}
errF <- learnErrors(filtFs, multithread=TRUE)

```

```{r}
errR <- learnErrors(filtRs, multithread=TRUE)

```

```{r}
dadaFs <- dada(filtFs, err=errF, multithread=TRUE)

```

```{r}
dadaRs <- dada(filtRs, err=errR, multithread=TRUE)

```

```{r}
mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=TRUE)
head(mergers[[1]])

```


```{r}
seqtab <- makeSequenceTable(mergers)
dim(seqtab)

```


```{r}
table(nchar(getSequences(seqtab)))

```

```{r}
seqtab2 <- seqtab[,nchar(colnames(seqtab)) %in% 250:256]
table(nchar(getSequences(seqtab2)))

```


```{r}
seqtab.nochim <- removeBimeraDenovo(seqtab2, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)


```


```{r}
print(dim(seqtab.nochim))
print(sum(seqtab.nochim))
sum(seqtab.nochim)/sum(seqtab2)

```


```{r}
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))

# If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)
colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
track %>% 
  as_tibble(rownames = "SampleID")

```


```{r}
track %>% 
  as_tibble(rownames = "SampleID") %>% 
  write_tsv("qc_tracked.tsv")

```

```{r, engine="bash"}
wget https://www2.decipher.codes/data/Downloads/TrainingSets/SILVA_SSU_r138_2_2024.RData

```

```{r}
dna <- DNAStringSet(getSequences(seqtab.nochim)) # Create a DNAStringSet from the ASVs
load("SILVA_SSU_r138_2_2024.RData") # CHANGE TO THE PATH OF YOUR TRAINING SET
ids <- IdTaxa(dna, trainingSet, strand="top", processors=NULL, verbose=FALSE)

```

```{r}
rownames(seqtab.nochim) <- sapply(strsplit(rownames(seqtab.nochim), "_"), `[`, 1)

asv_seqs <- colnames(seqtab.nochim)
asv_headers <- vector(dim(seqtab.nochim)[2], mode="character")
for (i in 1:dim(seqtab.nochim)[2]) {
  asv_headers[i] <- paste(">ASV", i, sep="_")
}

asv_fasta <- c(rbind(asv_headers, asv_seqs))
write(asv_fasta, "ASVs.fa")

```

```{r}

# tax table:
# creating table of taxonomy and setting any that are unclassified as "NA"
ranks <- c("domain", "phylum", "class", "order", "family", "genus")
asv_tax <- t(sapply(ids, function(x) {
  m <- match(ranks, x$rank)
  taxa <- x$taxon[m]
  taxa[startsWith(taxa, "unclassified_")] <- NA
  taxa
}))
colnames(asv_tax) <- ranks
rownames(asv_tax) <- gsub(pattern=">", replacement="", x=asv_headers)

asv_tax %>% 
  as_tibble(rownames = "ASV") %>% 
  count(domain)


asv_tax %>% 
  as_tibble(rownames = "ASV") %>% 
  write_tsv("ASVs_taxonomy.tsv")



```


```{r}
# count table:
asv_tab <- t(seqtab.nochim)
row.names(asv_tab) <- sub(">", "", asv_headers)
colnames(asv_tab) <- sample.names

asv_tab %>% 
  as_tibble(rownames = "ASV") %>% 
  write_tsv("ASVs_counts.tsv")

```



